Feat(api): Update API modules.

Derek-94 opened the pull request:
### Contents

- Make `responseInterface.ts` , that have **interfaces** related to api response.
- Update api modules.

### Details

1. `tsconfig.json` 관련

![image](https://user-images.githubusercontent.com/52649378/128103627-e2f082c7-f678-4e97-adcf-4f941aa209aa.png)

- 위 사진에 해당하는 친구들, `path` 관련 절대경로 사용하는 옵션들이 `compilerOptions` 외부에 선언되어 있어 위치를 옮겼습니다.

2. `types/responseInterface.ts` 작성

![image](https://user-images.githubusercontent.com/52649378/128103697-f36f5b02-6635-420a-9fd7-8a6654b57223.png)

- 래철님께서 제공해주신 `Swagger` 를 통해 정해진 `interface` 구현했습니다.

3. `api/api.ts` 작성

![image](https://user-images.githubusercontent.com/52649378/128104077-5f6f2fef-0ba6-48ad-a819-235bce16be0c.png)

- 위 `userAPI` 는 3개의 `method` 를 관리합니다. (`GET` , `POST`, `DELETE`, 추후 추가가능.)

- `GET`, `POST`, `DELETE` 메소드는 아래 함수로 구현했습니다.

![image](https://user-images.githubusercontent.com/52649378/128104266-fa42096e-2968-40c5-9aa9-2523687b8a1f.png)

---

### 궁금한 부분 🙋‍♂️

1. 제네릭 타입 관련

![image](https://user-images.githubusercontent.com/52649378/128104823-8573c4f9-f1a1-470d-9688-922aa976df1c.png)

- 사실 이부분을 제네릭을 꼭 써야했나, 생각이 듭니다.

- 고민했던 코드는 아래와 같습니다.

```typescript
const getRequest = async (url: string): Promise<any> => {
    try {
        const response = await axios.get(url);
        console.log(response);
        return response;
    } catch (e) {
        console.error(e.response);
        // return e.response;
    }
};
```

- 하지만 이는 `any` 타입이라 명시적이지 않아 적절치 않다 생각이 들었고, 또 다른 방법으로는

```typescript
const getRequest = async (url: string): Promise<UserInfo & UserGames & GameSlug & void> => {
    try {
        const response = await axios.get(url);
        console.log(response);
        return response;
    } catch (e) {
        console.error(e.response);
        // return e.response;
    }
};
```

- 이렇게 연산자를 통해서 다 명시해줘야 하나 생각들었지만, 명세가 업데이트 될때마다 수정해야하는 번거로움이 걱정되어 제네릭으로 선언했습니다.

- 이런 고민을 하다가.. 제너릭으로 구현했는데 다른분들의 의견이 궁금합니다.

2. 해당 3개 함수 내부 리턴값.

- `getRequest` 내부에서 `catch` 내부에서 개인적으로는 따로 리턴을 하지 않아서 이 부분이 괜찮으신지? 궁금합니다.

- `catch` 내부에서 반환이 되는 경우에는 `postRequest` 처럼 구현했습니다. 둘중 하나중 어떤게 좋을지 몰라 두가지로 구현했습니다. 

- 현업에서, 보통의 경우에서 `delete` method 수행시 리턴값이 필요한지, 에러처리를 위해 필요한지 궁금합니다!

---

🙋‍♂️ FE 분야에서 현업에서 일해본적은 없어서 어떤게 좋은 코드인지, 보편적인 코드인지 항상 궁금합니다.

코드리뷰 감사합니다!! 🙇‍♂️

closes #1

Derek-94 committed:
Feat(api): Update API modules.

- Make responseInterface.ts, that have interfaces related to api
  response.
- Update api modules.

kdongha commented on types/responseInterface.ts:
```diff
@@ -0,0 +1,21 @@
+export interface GameSlug {
```
GameSlug 보다는 Game이나 아래 UserInfo처러 GameInfo로 통일하는것이 좋지 않을까요?

Derek-94 replied:
넵 반영하겠습니다!

kdongha commented on api/api.ts:
```diff
@@ -0,0 +1,70 @@
+
+const END_POINT = 'http://localhost:8080';
+
+const getRequest = async <T>(url: string): Promise<T | void> => {
+    try {
+        const response = (await axios.get(url)) as T;
+        console.log(response);
+        return response;
+    } catch (e) {
+        console.error(e.response);
+        // return e.response;
+    }
+};
+
```
async로 감싼 이유가 있을까요?
따로 처리가 없다면 한번더 Promise로 감쌀 필요가 없을꺼 같아요.
또 END_POINT를 여기서 주입하는게 중복을 줄일수 있을꺼 같아요
```suggestion
const getRequest = <T>(url: string) => {
    try {
        return axios.get<T>(`${END_POINT}/${url}`));
    } catch (e) {
        console.error(e.response);
        // return e.response;
    }
};


```

Derek-94 replied:
넵 감사합니다.

궁금한점이 있는데요, 정상적으로 api 호출이 일어나지 않아 `catch` 문으로 들어올 때, 보통의 경우에선 어떤 프로세스를 거치는지 궁금합니다.

가령, 아무것도 없이 저렇게 `console` 를 출력한다던가, 혹은 에러값 `e` 도 반환을 한다던가, 어떤게 좋은 방법인지 궁금합니다.

kdongha reviewed (commented):
delete의 경우에는 상태코드에따라 return값을 쓸지 말지 결정하면 될거 같습니다~

Derek-94 committed:
feat(feedback): Applied feedback.

- Delted async await in Request functions.
- Put END_POINT in Request functions.

Derek-94 commented:
@kdongha 동하님, 피드백 적용했습니다.

다만, `deleteRequest` 함수 내부에서 status 에 접근할 수 가 없어 고민입니다.

![image](https://user-images.githubusercontent.com/52649378/128651368-bb19791f-96b4-4a6f-b185-d88bffa6e003.png)

위 코드를 

```typescript
const deleteRequest = (url: string) => {
    try {
        const response = axios.delete(`${END_POINT}/${url}`);
        console.log(response);
        if(response.status === 204) {
          // 정상 처리 코드
        }
        // return response;
    } catch (e) {
        console.error(e);
    }
};
```

이런식으로 수정하려했는데, 이 함수에 중첩된 `async await` 를 뺐더니 안되더라구요. 

혹시 이것 관련해서 `deleteRequest ` 를 `async await` 로 감싸지 않고 접근할 수 있는 법이 없지 않나 싶습니다. 😥 

또한 중첩된? `async await` 에는 (제가 처음에 구현한 내용) 어떤 부작용? 이 있는지 궁금합니다~!

kdongha commented on api/api.ts:
```diff
@@ -0,0 +1,66 @@
+    }
+};
+
+const deleteRequest = (url: string) => {
+    try {
+        const response = axios.delete(`${END_POINT}/${url}`);
+        console.log(response);
+        // return response;
+    } catch (e) {
+        console.error(e);
+    }
+};
```
제가 생각했던 getRequest, deleteRequest, postRequest의 경우 나중에 axios가 아닌 fetch로 변경한다던가 할 경우에 전체 코드 변경하지 않고 해당 부분만 변경해서 적용하는 용으로 쓰고 싶었습니다.
```suggestion
const deleteRequest = (url: string) => {
    return axios.delete(`${END_POINT}/${url}`);
};

```
그래서 이렇게 그냥 처리하고 각각 상황에 따른 처리는 사용하는 쪽에서 처리하려고 했습니다.

Derek-94 replied:
앗 이해했습니다! 해당 모듈 사용하는 부분에서 사용한다는 의미군요! 감사합니다 🙏

kdongha commented on api/api.ts:
```diff
@@ -0,0 +1,66 @@
+const postRequest = <T>(url: string, body: any) => {
+    try {
+        const response = axios.post<T>(`${END_POINT}/${url}`, body);
+        console.log(response);
```
이경우엔 콘솔에 Promise밖에 안찍힐것 같습니다.

Derek-94 replied:
넵! 삭제하겠습니다.

kdongha reviewed (commented):
중첩된 async await의 경우 부작용은 없지만 필요없는 코드가 한번 더 들어갔다고 생각되었습니다.

Derek-94 committed:
feat(api): Revised delete module.

kdongha commented on api/api.ts:
```diff
@@ -16,7 +16,7 @@ const deleteRequest = (url: string) => {
     try {
         const response = axios.delete(`${END_POINT}/${url}`);
         console.log(response);
```
여기두 마찬가지로 promise밖에 안찍힙니다.

Derek-94 replied:
앗 넵 삭제하겠습니닷

Derek-94 committed:
feat(api): Delete unnecessary console log.
